akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -495,12 +495,14 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
       case LoadSnapshotFailed(cause) ⇒
         timeoutCancellable.cancel()
         try onRecoveryFailure(cause, event = None) finally context.stop(self)
+        returnRecoveryPermit()
 
       case RecoveryTick(true) ⇒
         try onRecoveryFailure(
           new RecoveryTimedOut(s"Recovery timed out, didn't get snapshot within $timeout"),
           event = None)
         finally context.stop(self)
+        returnRecoveryPermit()
```

Comment:
should those be in the finally blocks? in case a class implements onRecoveryFailure and ends up throwing for some reason, we still want to return the token.

Great catch btw, I'll make sure to have those covered in the typed one

Reply:
There is a catch surrounding the whole message processing that will handle that case. In typed I assume this will be a lot easier because we don't have the restriction of not being able to introduce new state variables as we have in the Eventsourced trait.

Reply:
Ah good, yeah indeed somewhat messy. Thanks!